File: history/0.1.2/tatala-client-csharp/src/QiLeYuan/Tatala/socket/client/ShortClientSession.cs
using System;
using System.IO;
using System.Net;
using System.Net.Sockets;
using QiLeYuan.Tatala.socket.to;
using QiLeYuan.Tatala.socket.util;
using QiLeYuan.Tools.debug;

namespace QiLeYuan.Tatala.socket.client {
    public class ShortClientSession {

        private String hostIp;
        private int hostPort;
        private int timeout;
        private int retryTime;

        // State object for receiving data from remote device.
        public class StateObject {
            // Client socket.
            public Socket workSocket = null;
            // Size of receive buffer.
            public const int BufferSize = 256;
            // Receive buffer.
            public byte[] buffer = new byte[BufferSize];
            // TransferObject
            public TransferObject transferObject = null;
            // Received data
            public MemoryStream memoryStream = new MemoryStream();
        }

        public ShortClientSession(String hostIp, int hostPort, int timeout, int retryTime) {
            this.hostIp = hostIp;
            this.hostPort = hostPort;
            this.timeout = timeout;
            this.retryTime = retryTime;
        }

        public Object start(TransferObject to) {
            Object resultObject = null;

            String calleeClass = to.getCalleeClass();
            String calleeMethod = to.getCalleeMethod();

            StateObject stateObject = new StateObject();
            stateObject.transferObject = to;

            try {
                if (Connect(stateObject)) {

                    Send(stateObject);

                    resultObject = Receive(stateObject);
                }
            } catch (Exception e) {
                Logging.LogError("Callee Class and Method: [" + calleeClass + "." + calleeMethod + "]");
                Logging.LogError(e.ToString());
            } finally {
                Close(stateObject);
            }

            return resultObject;
        }

        private bool Connect(StateObject stateObject) {
            Socket client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
            IPEndPoint ipEndPoint = new IPEndPoint(IPAddress.Parse(hostIp), hostPort);

            stateObject.workSocket = client;

            int retry = retryTime;
            while (true) {
                try {
                    // Connect to the remote endpoint.
                    client.Connect(ipEndPoint);
                    break;
                } catch (Exception be) {
                    retry--;
                    Logging.LogError(be.Message);
                    Logging.LogError("Retry time: " + retry);
                }
                if (retry <= 0) {
                    Logging.LogError("Connecntion Error.");
                    return false;
                }
            }

            client.SendTimeout = timeout;
            client.ReceiveTimeout = timeout;

            return true;
        }

        private void Send(StateObject stateObject) {
            TransferObject to = stateObject.transferObject;
            Socket client = stateObject.workSocket;

            byte[] sendData = TransferUtil.transferObjectToByteArray(to);

            client.Send(sendData);
        }

        private Object Receive(StateObject stateObject) {
            Object resultObject = null;

            //if return void, don't call socket receive
            if (stateObject.transferObject.getReturnType() == TransferObject.DATATYPE_VOID) {
                return resultObject;
            }

            Socket client = stateObject.workSocket;
            int receiveLength = 0;
            int expectReceiveLength = 0;
            bool firsttime = true;
            while (true) {
                int bytesReceived = client.Receive(stateObject.buffer, 0, StateObject.BufferSize, SocketFlags.None);

                stateObject.memoryStream.Write(stateObject.buffer, 0, bytesReceived);
                receiveLength += bytesReceived;

                //check whether one server response is done or not
                if (bytesReceived < StateObject.BufferSize) {
                    break;
                }
                if (bytesReceived == StateObject.BufferSize) {
                    if (firsttime) {
                        expectReceiveLength = TransferUtil.getExpectReceiveLength(stateObject.buffer);
                    }
                    //if expectReceiveLength equal receiveLength, the server response is done
                    if (expectReceiveLength == receiveLength) {
                        break;
                    }
                }
                firsttime = false;
            }

            byte[] receiveData = stateObject.memoryStream.ToArray();
            resultObject = TransferUtil.byteArrayToReturnObject(receiveData);

            return resultObject;
        }

        private void Close(StateObject stateObject) {
            stateObject.memoryStream.Close();
            if (stateObject.workSocket.Connected) {
                stateObject.workSocket.Shutdown(SocketShutdown.Both);
            }
            stateObject.workSocket.Close();
        }
    }
}

File: tatala-java/src/com/qileyuan/tatala/socket/client/ShortClientSession.java
package com.qileyuan.tatala.socket.client;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.BindException;
import java.net.Socket;
import java.net.SocketException;
import java.net.SocketTimeoutException;
import java.util.zip.DataFormatException;
import java.util.zip.Inflater;

import org.apache.log4j.Logger;

import com.qileyuan.tatala.socket.exception.SocketExecuteException;
import com.qileyuan.tatala.socket.to.TransferObject;
import com.qileyuan.tatala.socket.util.TransferUtil;

public class ShortClientSession{
	static Logger log = Logger.getLogger(ShortClientSession.class);
	
	private String hostIp;
	private int hostPort;
	private int timeout;
	private int retryTime;
	
	public ShortClientSession(String hostIp, int hostPort, int timeout, int retryTime){
		this.hostIp = hostIp;
		this.hostPort = hostPort;
		this.timeout = timeout;
		this.retryTime = retryTime;
	}
	
	public Object start(TransferObject to) {
		Object resultObject = null;
		Socket client = null;
		String calleeClass = to.getCalleeClass();
		String calleeMethod = to.getCalleeMethod();
		try {
			if(client == null){
				client = connect();
			}

			send(client, to);
			
			resultObject = receive(client, to);
		} catch (BindException be) {
			log.error("Connection error: " + be.getMessage());
			log.error("Callee Class and Method: [" + calleeClass + "." + calleeMethod + "]");
		} catch (SocketTimeoutException ste) {
			log.error("Socekt timed out, return null. [" + timeout + "ms]");
			log.error("Callee Class and Method: [" + calleeClass + "." + calleeMethod + "]");
		} catch (Exception e) {
			log.error("Callee Class and Method: [" + calleeClass + "." + calleeMethod + "]");
			e.printStackTrace();
		} finally {
			try {
				if (client != null) {
					close(client);
				}
			} catch (IOException e) {}
		}

		return resultObject;
	}

	private Socket connect() throws SocketException {
		Socket client = null;
		String errorMessage = "";
		int retry = retryTime;
		while (client == null && retry > 0) {
			try {
				client = new Socket(hostIp, hostPort);
			} catch (Exception e) {
				log.error("Connection error: " + e.getMessage());
				errorMessage = e.getMessage();
			}
			retry--;
			if (client == null) {
				log.error("Retry time: " + retry);
			}
		}
		if (client == null) {
			throw new BindException(errorMessage);
		}

		client.setSoTimeout(timeout);
		client.setReuseAddress(true);
		//this can avoid socket TIME_WAIT state
		//client.setSoLinger(true, 0);

		return client;
	}

	private void send(Socket client, TransferObject to) throws IOException {
		OutputStream os = client.getOutputStream();

		byte[] sendData = TransferUtil.transferObjectToByteArray(to);
		
		os.write(sendData);
	}

	private Object receive(Socket client, TransferObject to) throws IOException, DataFormatException, SocketExecuteException {
		Object resultObject = null;

		InputStream is = client.getInputStream();
		// in
		int compressFlag = is.read();
		if (compressFlag == 1) {
			resultObject = doInCompress(is);
		} else {
			resultObject = doInNormal(is);
		}

		return resultObject;
	}

	private void close(Socket client) throws IOException {
		OutputStream os = client.getOutputStream();
		InputStream is = client.getInputStream();
		os.close();
		is.close();
		client.close();
	}

	private Object doInNormal(InputStream is) throws IOException, SocketExecuteException {
		return TransferObject.convertReturnInputStream(is);
	}
	
	private Object doInCompress(InputStream is) throws IOException,	DataFormatException, SocketExecuteException {
		// in
		int unCompressedLength = TransferUtil.readInt(is);
		int compressedLength = TransferUtil.readInt(is);
		byte[] input = new byte[compressedLength];
		is.read(input);
		byte[] output = new byte[unCompressedLength];
		Inflater decompresser = new Inflater();
		decompresser.setInput(input);
		decompresser.inflate(output);
		decompresser.end();

		InputStream istemp = new ByteArrayInputStream(output);
		return TransferObject.convertReturnInputStream(istemp);
	}
}
